src/libcollections/vec.rs

```diff
@@ -1387,6 +1387,17 @@ impl<T> FromIterator<T> for Vec<T> {
         let (lower, _) = iterator.size_hint();
         let mut vector = Vec::with_capacity(lower);
 
+        // This function should be the moral equivalent of:
+        //
+        //      for item in iterator {
+        //          vector.push(item);
+        //      }
+        //
+        // This equivalent crucially runs the iterator precisely once. The
+        // optimization below (eliding bound/growth checks) means that we
+        // actually run the iterator twice. To ensure the "moral equivalent" we
+        // do a `fuse()` operation to ensure that the iterator continues to
+        // return `None` after seeing the first `None`.
         let mut i = iterator.fuse();
```

Comment:
FWIW, I'm not super keen on this `fuse` approach. In general it adds a check to every `next` call, to protect against a case that is "unspecified behaviour" anyway (the iterator being shorter than its stated `size_hint`).

I'd prefer adding a check after the loop:

```rust
if vector.len() < vector.capacity() { return vector }
```

since that is guaranteed to be only a single check and outside the inner loop, even if the optimiser can't remove it.